File: server/common/communication.py
BUFSIZE = 1024
SIZE_I32 = 4
ENDIAN_ORDER = 'big'
STRING_ENCODING = 'ascii'

def send_bytes(socket, msg: bytes):
    total_sent = 0
    while total_sent < len(msg):
        sent = socket.send(msg[total_sent:]) 
        if sent == 0:
            raise RuntimeError("socket send error")
        total_sent += sent 

def recv_bytes(socket, size: int) -> bytes:
    data_received = b''
    while len(data_received) < size:
        chunk = socket.recv(min(size - len(data_received), BUFSIZE))
        if len(chunk) == 0:
            raise RuntimeError("socket recv error")
        data_received += chunk
    return data_received

def send_i32(socket, n: int):
    data = n.to_bytes(SIZE_I32, byteorder=ENDIAN_ORDER, signed=True)
    send_bytes(socket, data)

def recv_i32(socket) -> int:
    data = recv_bytes(socket, SIZE_I32)
    return int.from_bytes(data, byteorder=ENDIAN_ORDER, signed=True)

def send_string(socket, s: str):
    data = bytes(s, encoding=STRING_ENCODING)
    send_bytes(socket, data)

def recv_string(socket, size: int) -> str:
    data = recv_bytes(socket, size)
    return data.decode(str=STRING_ENCODING)

File: client/common/communication.go
package common

import (
	"encoding/binary"
	"errors"
	"fmt"
	"net"
)

const SIZE_U32 = 4

func sendBytes(socket net.Conn, msg []byte) error {
	totalSent := 0
	for totalSent < len(msg) {
		sent, err := socket.Write(msg[totalSent:])
		if err != nil {
			return errors.New("socket send error")
		}
		totalSent += sent
	}
	return nil
}

func recvBytes(socket net.Conn, size uint32) ([]byte, error) {
	buffer := make([]byte, size)
	var dataReceivedCounter uint32 = 0
	for dataReceivedCounter < size {
		n, err := socket.Read(buffer[dataReceivedCounter:])
		if err != nil {
			return buffer, errors.New("socket recv error")
		}
		dataReceivedCounter += uint32(n)
	}
	return buffer, nil
}

func sendU32(socket net.Conn, n uint32) error {
	bs := make([]byte, SIZE_U32)
	binary.BigEndian.PutUint32(bs, n)
	return sendBytes(socket, bs)
}

func recvU32(socket net.Conn) (uint32, error) {
	data, err := recvBytes(socket, SIZE_U32)
	if err != nil {
		return 0, err
	}
	n := binary.BigEndian.Uint32(data)
	return n, nil
}

func sendString(socket net.Conn, s string) error {
	size := uint32(len(s))
	err := sendU32(socket, size)
	if err != nil {
		return err
	}

	data := []byte(s)
	err = sendBytes(socket, data)
	if err != nil {
		return err
	}

	return nil
}

func recvString(socket net.Conn) (string, error) {
	size, err := recvU32(socket)
	if err != nil {
		return "", err
	}

	data, err := recvBytes(socket, size)
	if err != nil {
		return "", err
	}

	return string(data[:]), nil
}

func sendBet(socket net.Conn, name string, surname string, document string, birthdate string, number int) error {
	res := fmt.Sprintf("%v,%v,%v,%v,%v", name, surname, document, birthdate, number)
	return sendString(socket, res)
}
